semgrep-core/src/matching/Generic_vs_generic.ml

```diff
@@ -446,136 +454,117 @@ let rec m_name a b =
+             * > different.
+             *
+             * FIXME:
+             * This is actually not the correct way of dealing with the problem,
+             * because there could be `metavariable-xyz` operators filtering the
+             * potential values of the metavariable. See DeepSemgrep commit
+             *
+             *     5b2766ee30e "test: Tests for matching metavariable patterns against resolved names"
+          *)
+          | G.Id ((str, _tok), _info) when MV.is_metavar_name str -> fail ()
+          | _ ->
+              (* Try matching against parent classes *)
+              try_parents dotted)
+      | __else__ -> fail ())
+  | a, B.IdQualified nameinfo -> (
       (* try without resolving anything *)
-      >||> m_name a
-             (B.IdQualified { nameinfo with name_info = B.empty_id_info () })
-      >||>
-      (* try this time by replacing the qualifier by the resolved one *)
-      let new_qualifier =
-        match List.rev dotted with
-        | [] -> raise Impossible
-        | _x :: xs -> List.rev xs |> Common.map (fun id -> (id, None))
-      in
-      m_name a
-        (B.IdQualified
-           {
-             nameinfo with
-             name_middle = Some (B.QDots new_qualifier);
-             name_info = B.empty_id_info ();
-           })
-  (* semantic! try to handle open in OCaml by querying LSP! The
-   * target code is using an unqualified Id possibly because of some open!
-   *)
-  | G.IdQualified { name_last = ida, None; _ }, B.Id (idb, _infob)
-    when fst ida = fst idb -> (
-      match !Hooks.get_def idb with
-      | None -> fail ()
-      | Some file ->
-          let m = module_name_of_filename file in
-          let t = snd idb in
-          pr2_gen m;
-          let _n = H.name_of_ids [ (m, t); idb ] in
-          (* retry with qualified target *)
-          (* m_name a n *)
-          return ())
-  (* boilerplate *)
-  | ( _,
-      G.IdQualified
-        ({
-           name_info =
-             {
-               B.id_resolved =
-                 {
-                   contents =
-                     Some
-                       ( (( B.ImportedEntity dotted
-                          | B.ImportedModule (B.DottedName dotted)
-                          | B.ResolvedName (dotted, _) ) as resolved),
-                         _sid );
-                 };
-               _;
-             };
-           _;
-         } as b1) ) -> (
-      try_parents dotted
-      >||> try_alternate_names resolved
-      >||>
-      match a with
-      | IdQualified a1 -> m_name_info a1 b1
-      | Id _ -> fail ())
-  | G.IdQualified a1, B.IdQualified b1 -> m_name_info a1 b1
-  | G.Id _, _
-  | G.IdQualified _, _ ->
-      fail ()
+      (match a with
+      | B.Id ((str, tok), _info) when MV.is_metavar_name str ->
+          (* If `is_distributed` is true, then we got the target `IdQualified` from a
+             `ResolvedName`.
```

Comment:
Is this true? Doesn't this also handle the case where the `IdQualified` comes from `Naming_AST`?